app/src/main/java/pm/gnosis/heimdall/ui/transactions/details/GenericTransactionDetailsFragment.kt

```diff
@@ -0,0 +1,166 @@
+    private val inputSubject = PublishSubject.create<GenericTransactionDetailsContract.CombinedRawInput>()
+    private var editable: Boolean = false
+    private var originalTransaction: Transaction? = null
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View?
+            = inflater.inflate(R.layout.layout_transaction_details_generic, container, false)
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        editable = arguments?.getBoolean(ARG_EDITABLE, false) ?: false
+        val transaction = arguments?.getParcelable<TransactionParcelable>(ARG_TRANSACTION)?.transaction
+        originalTransaction = transaction
+        layout_transaction_details_generic_to_input.setText(transaction?.address?.asEthereumAddressStringOrNull())
+        layout_transaction_details_generic_data_input.setText(transaction?.data)
+        layout_transaction_details_generic_value_input.setText(transaction?.value?.value?.asDecimalString())
+        toggleTransactionInput(editable)
+    }
+
+    override fun onStart() {
+        super.onStart()
+        val safe = arguments?.getString(ARG_SAFE)?.hexAsBigIntegerOrNull()
+        setupSafeSpinner(layout_transaction_details_generic_safe_input, safe)
+        disposables += Observable.combineLatest(
+                prepareInput(layout_transaction_details_generic_to_input),
+                prepareInput(layout_transaction_details_generic_value_input),
+                prepareInput(layout_transaction_details_generic_data_input),
+                Function3 { to: CharSequence, value: CharSequence, data: CharSequence ->
+                    GenericTransactionDetailsContract.CombinedRawInput(to.toString() to false, value.toString() to false, data.toString() to false)
+                }
+        ).subscribe(inputSubject::onNext, Timber::e)
+    }
+
+    override fun inputEnabled(enabled: Boolean) {
+        if (editable) {
+            toggleTransactionInput(enabled)
+        }
+    }
+
+    private fun toggleTransactionInput(enabled: Boolean) {
+        layout_transaction_details_generic_safe_input.isEnabled = enabled
+        layout_transaction_details_generic_to_input.isEnabled = enabled
+        layout_transaction_details_generic_data_input.isEnabled = enabled
+        layout_transaction_details_generic_value_input.isEnabled = enabled
+    }
+
+    override fun selectedSafeChanged(safe: Safe?) {
+        safeSubject.onNext(safe?.address?.toOptional() ?: None)
+    }
+
+    override fun observeTransaction(): Observable<Result<Transaction>> {
+        return inputSubject
+                .compose(subViewModel.inputTransformer(context!!, originalTransaction))
+                .observeOn(AndroidSchedulers.mainThread())
+                .doOnNext {
+                    (it as? ErrorResult)?.let {
+                        (it.error as? TransactionInputException)?.let {
```

Comment:
If we want to spare a line here:
`((it as? ErrorResult)?.error as? TransactionInputException)?.let  {...}`